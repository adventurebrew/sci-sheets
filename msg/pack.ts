import {MsgData, MsgResource, TranslationLine} from "../types/SCI";
import {
  writeString,
  writeUint16,
  writeUint32,
  writeUint8
} from "../utils/binary";
import {sumOf} from "../utils/collections";
import {PackOptions, SIZE_ENTRY, SIZE_HEADER} from "../common";

const getZeroTerminatedStringLength = (line: string) => line.length + 1;

export function packTranslation(lines: TranslationLine[], options: PackOptions) {
  const msgResource = translation2msg(lines);
  return packMSG(msgResource, options);
}

export function translation2msg(lines: TranslationLine[]): MsgResource {
  const strings = lines.map(l => l.text);
  const data: MsgData = {
    version: 0xD0C008F,
    length: SIZE_HEADER
      + (lines.length * SIZE_ENTRY)
      + sumOf(strings, getZeroTerminatedStringLength),
    lastMsgNum: 0,
    nEntries: lines.length,
    entries: [],
  };

  let stringsOffset = SIZE_HEADER + (lines.length * SIZE_ENTRY);
  for (const line of lines) {
    data.entries.push({
      noun: line.noun,
      verb: line.verb,
      condition: line.condition,
      sequence: line.sequence,
      talker: line.talker,
      offset: stringsOffset - 2,
      refNoun: line.refNoun,
      refVerb: line.refVerb,
      refCond: line.refCond,
      refSeq: 0,
    });

    stringsOffset += line.text.length + 1;
  }

  return { data, strings };
}

export function packMSG(msg: MsgResource, options: PackOptions): Uint8Array {
  let index: number;
  let entryOffset: number;

  const nEntries = msg.data.nEntries;
  const fileSize = SIZE_HEADER
    + (nEntries * SIZE_ENTRY)
    + sumOf(msg.strings, getZeroTerminatedStringLength);

  const output = new Uint8Array(fileSize);
  writeUint32(output, 0, 0xD0C008F);
  writeUint16(output, 6, output.length);
  writeUint16(output, 8, msg.strings.length);

  entryOffset = SIZE_HEADER;
  for (index = 0; index < nEntries; index++) {
    const entry = msg.data.entries[index];
    writeUint8(output,  entryOffset + 0, entry.noun);
    writeUint8(output,  entryOffset + 1, entry.verb);
    writeUint8(output,  entryOffset + 2, entry.condition);
    writeUint8(output,  entryOffset + 3, entry.sequence);
    writeUint8(output,  entryOffset + 4, entry.talker);
    writeUint16(output, entryOffset + 5, entry.offset);
    writeUint8(output,  entryOffset + 7, entry.refNoun);
    writeUint8(output,  entryOffset + 8, entry.refVerb);
    writeUint8(output,  entryOffset + 9, entry.refCond);
    writeString(options.encoding, output, entry.offset + 2, msg.strings[index]);

    entryOffset += SIZE_ENTRY;
  }

  return output;
}

