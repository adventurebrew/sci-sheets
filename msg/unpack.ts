import {IndexEntry} from "../types/SCI";
import {readUint16, readUint8} from "../utils/binary";

export function readEntry(buffer: Uint8Array, index: number): IndexEntry {
    const entryOffset = (1 + index) * 10;

    const noun      = readUint8(buffer, entryOffset + 0);
    const verb      = readUint8(buffer, entryOffset + 1);
    const condition = readUint8(buffer, entryOffset + 2);
    const sequence  = readUint8(buffer, entryOffset + 3);
    const talker    = readUint8(buffer, entryOffset + 4);
    const offset    = readUint16(buffer, entryOffset + 5);
    const refNoun   = readUint8(buffer, entryOffset + 7);
    const refVerb   = readUint8(buffer, entryOffset + 8);
    const refCond   = readUint8(buffer, entryOffset + 9);
    const refSeq    = readUint8(buffer, entryOffset + 10);

    return {
        noun,
        verb,
        condition,
        sequence,
        talker,
        offset,
        refNoun,
        refVerb,
        refCond,
        refSeq,
    };
}
