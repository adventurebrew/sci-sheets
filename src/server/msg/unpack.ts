import {IndexEntry, MsgData, MsgResource, TranslationLine} from '../types/SCI';
import {readString, readUint16, readUint32, readUint8} from '../utils/binary';
import {PackOptions, SIZE_ENTRY, SIZE_HEADER} from './common';

export function unpackTranslation(buffer: Uint8Array, options: PackOptions) {
  return msg2translation(unpackMSG(buffer, options));
}

export function unpackMSG(buffer: Uint8Array, options: PackOptions): MsgResource {
  let index: number;
  let entry: IndexEntry;

  const { encoding } = options;
  const data = readMsgData(buffer);
  const strings = new Array(data.nEntries);

  for (index = 0; index < data.nEntries; index++) {
    entry = data.entries[index];
    strings[index] = readString(encoding, buffer, entry.offset + 2);
  }

  return { data, strings };
}

export function msg2translation(msg: MsgResource): TranslationLine[] {
  let index: number;
  let entry: IndexEntry;

  const { nEntries } = msg.data;
  const lines = new Array<TranslationLine>(nEntries);

  for (index = 0; index < nEntries; index++) {
    entry = msg.data.entries[index];

    lines[index] = {
      noun: entry.noun,
      verb: entry.verb,
      condition: entry.condition,
      sequence: entry.sequence,
      talker: entry.talker,
      refNoun: entry.refNoun,
      refVerb: entry.refVerb,
      refCond: entry.refCond,
      text: msg.strings[index]
    };
  }

  return lines;
}


function readMsgData(buffer: Uint8Array): MsgData {
  const version = readUint32(buffer, 0);
  const length = readUint16(buffer, 4);
  const lastMsgNum = readUint16(buffer, 6);
  const nEntries = readUint16(buffer, 8);
  const entries = new Array(nEntries);

  for (let index = 0; index < nEntries; index++) {
    entries[index] = readEntry(buffer, index);
  }

  return {
    version,
    length,
    lastMsgNum,
    nEntries,
    entries,
  };
}

function readEntry(buffer: Uint8Array, index: number): IndexEntry {
  const entryOffset = SIZE_HEADER + (SIZE_ENTRY * index);

  const noun      = readUint8(buffer, entryOffset + 0);
  const verb      = readUint8(buffer, entryOffset + 1);
  const condition = readUint8(buffer, entryOffset + 2);
  const sequence  = readUint8(buffer, entryOffset + 3);
  const talker    = readUint8(buffer, entryOffset + 4);
  const offset    = readUint16(buffer, entryOffset + 5);
  const refNoun   = readUint8(buffer, entryOffset + 7);
  const refVerb   = readUint8(buffer, entryOffset + 8);
  const refCond   = readUint8(buffer, entryOffset + 9);
  const refSeq    = 0;

  return {
    noun,
    verb,
    condition,
    sequence,
    talker,
    offset,
    refNoun,
    refVerb,
    refCond,
    refSeq,
  };
}
