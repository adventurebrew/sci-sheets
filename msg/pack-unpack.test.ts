import {unpackMSG, unpackTranslation} from './unpack';
import {packMSG, packTranslation} from "./pack";

describe('pack-unpack test', () => {
  const MSG95 = Buffer.from(
    `jwAMDQAAvgUWAAEBAAFj5AAUALoCAQABY/QApqamBAEAAWMhAaampgUEAAFjTgF3AAAGBAABY2wB` +
    `APR3BgQBAWO6AaampgYBAAFj3wEAAAAHAQABYwYCAAAACAQAAWM0AqampgkBAAFjWAJBU1QKAAIB` +
    `Y4YCAAIACgADAWOjAnempgoABAFjugKmNacKAAUBYykDwncACgEAAWNTA6bCdwozAAFjFASmpqYL` +
    `BAABYzIEAAAACwEAAWOWBHeVdwwBAAFj8wSmpqYDBAABY04FAAAAAwQBAWNzBQCmpgMBAAFjmAWm` +
    `pqZUaGUgYmFjayBkb29yLiAAVGhleSByZW1pbmQgeW91IG9mIGEgYnVuY2ggb2YgS25vY2hlbGtv` +
    `cGZzLiAAVGhlcmUgaXMgYSBjaGFuZGVsaWVyIHdpdGggYSByb3BlIGF0dGFjaGVkLiAAVGhlIGRv` +
    `b3IgaXMgYWxyZWFkeSBibG9ja2VkLiAATm8gdGltZSBmb3IgdGhhdCBhZ2FpbiAoYWx0aG91Z2gg` +
    `aXQgV0FTIGZ1bikuICBKdXN0IGdldCB0aGUgaGVjayBvdXR0YSBoZXJlLiAAVGhlIGNoYW5kZWxp` +
    `ZXIgaXMgb3V0IG9mIHlvdXIgcmVhY2guAENoYW5kZWxpZXIsIHVzZWQgdG8gaG9sZCB1cCB0aGUg` +
    `cm9wZS4gAEhlJ3MgYmVlbiBkcmlua2luZyBUcm9sbCdzIHN3ZWF0IGFsbCBuaWdodC4gIABUaGUg` +
    `ZnJvbnQgZG9vciBpcyBhbHJlYWR5IGJhcnJlZC4gIABUaGUgZmxvb3IgaXMgc3RyZXduIHdpdGgg` +
    `ZmxvdHNhbSBhbmQgamV0c2FtLiAAWW91IGNsb3NlIGFuZCBiYXIgdGhlIGRvb3IuIAAiT2gsIGEg` +
    `d2lzZSBndXksIGVoPyIgAFRoZXJlJ3Mgbm8gbmVlZCB0byBrbm9jayB0aGlzIGd1eSBvdXQ7IHRo` +
    `ZSBUcm9sbCdzIFN3ZWF0IGhlJ3MgYmVlbiBkcmlua2luZyBoYXMgYWxyZWFkeSBkb25lIHRoZSBq` +
    `b2IgZm9yIHlvdS4gAFlvdSBtb3ZlIHRoZSBjaGFpciBpbiBmcm9udCBvZiB0aGUgZG9vci4gAFdv` +
    `b2RlbiB0YWJsZXMsIGJlbmNoZXMsIGFuZCBjaGFpcnMgZmlsbCB0aGUgc3BhY2Ugb2YgdGhpcyBi` +
    `cmlnYW5kIG1lZXRpbmcgYXJlYS4gIEEgY2hhbmRlbGllciBsaWdodHMgdGhlIHJvb20gYW5kICBh` +
    `biB1bmxpdCBjYW5kZWxhYnJhIHN0YW5kcyBpbiB0aGUgY29ybmVyLiAgQSByb3BlIGhhbmdzIGZy` +
    `b20gdGhlIGNoYW5kZWxpZXIuIABOb25lIG9mIHRoZSBkb29ycyBhcmUgbG9ja2VkLgBMZWF2ZSB0` +
    `aGUgcm9wZSBhbG9uZSBmb3Igbm93LiAgSXQncyBzZXJ2aW5nIGEgdmVyeSBpbXBvcnRhbnQgZnVu` +
    `Y3Rpb24gYnkgaG9sZGluZyB1cCB0aGUgY2hhbmRlbGllci4AVGhlIHJvcGUgaG9sZHMgdXAgdGhl` +
    `IGNoYW5kZWxpZXIuICBJdCdzIHVzZWQgdG8gbG93ZXIgdGhlIGNoYW5kZWxpZXIgc28gdGhhdCBp` +
    `dCBjYW4gYmUgbGl0LiAATW9zdCBvZiB0aGUgdGFibGVzIGFyZSBncmlteSB3aXRoIGRyaWVkIGJl` +
    `ZXIgYW5kIHN1Y2guICBUaGUgaGVhZCB0YWJsZSwgdGhvdWdoLCBpcyBjbGVhbi4gAFlvdSBiZWdp` +
    `biB0b3BwbGluZyB0aGUgY2FuZGVsYWJyYS4gIABUaGUgY2FuZGVsYWJyYSBoYXMgYWxyZWFkeSBm` +
    `YWxsZW4uICAAU3R1cmR5LWxvb2tpbmcgZGV2aWNlIGZvciBob2xkaW5nIGNhbmRsZXMuIAAAAAAA` +
    `AAAAAAAAAAAAAAAAAAAAAAAA`, 'base64'
  );

  it('should unpack a message resource file', () => {
    const msg = unpackMSG(MSG95, { encoding: 'ascii' });
    expect(msg).toMatchSnapshot();
  });

  it('should pack a message resource file back', () => {
    const msg = unpackMSG(MSG95, { encoding: 'ascii' });
    const buffer = packMSG(msg, { encoding: 'ascii' });
    expect(Buffer.from(buffer)).toEqual(MSG95);
  });

  it('should convert a message resource file to translations', () => {
    const translation = unpackTranslation(MSG95, { encoding: 'ascii' });
    expect(translation).toMatchSnapshot();
  });

  it('should convert a message resource to translations and back again', () => {
    const translation = unpackTranslation(MSG95, { encoding: 'ascii' });
    const buffer = packTranslation(translation, { encoding: 'ascii' });

    expect(Buffer.from(buffer)).toEqual(MSG95);
  });
});
