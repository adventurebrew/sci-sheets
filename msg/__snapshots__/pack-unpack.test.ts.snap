// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pack-unpack test should convert a message resource file to translations 1`] = `
Array [
  Object {
    "condition": 0,
    "index": 0,
    "noun": 1,
    "refCond": 186,
    "refNoun": 20,
    "refVerb": 0,
    "sequence": 1,
    "talker": 99,
    "text": "The back door. ",
    "verb": 1,
  },
  Object {
    "condition": 0,
    "index": 1,
    "noun": 2,
    "refCond": 166,
    "refNoun": 166,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "They remind you of a bunch of Knochelkopfs. ",
    "verb": 1,
  },
  Object {
    "condition": 0,
    "index": 2,
    "noun": 4,
    "refCond": 166,
    "refNoun": 166,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "There is a chandelier with a rope attached. ",
    "verb": 1,
  },
  Object {
    "condition": 0,
    "index": 3,
    "noun": 5,
    "refCond": 0,
    "refNoun": 119,
    "refVerb": 0,
    "sequence": 1,
    "talker": 99,
    "text": "The door is already blocked. ",
    "verb": 4,
  },
  Object {
    "condition": 0,
    "index": 4,
    "noun": 6,
    "refCond": 119,
    "refNoun": 0,
    "refVerb": 244,
    "sequence": 1,
    "talker": 99,
    "text": "No time for that again (although it WAS fun).  Just get the heck outta here. ",
    "verb": 4,
  },
  Object {
    "condition": 1,
    "index": 5,
    "noun": 6,
    "refCond": 166,
    "refNoun": 166,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "The chandelier is out of your reach.",
    "verb": 4,
  },
  Object {
    "condition": 0,
    "index": 6,
    "noun": 6,
    "refCond": 0,
    "refNoun": 0,
    "refVerb": 0,
    "sequence": 1,
    "talker": 99,
    "text": "Chandelier, used to hold up the rope. ",
    "verb": 1,
  },
  Object {
    "condition": 0,
    "index": 7,
    "noun": 7,
    "refCond": 0,
    "refNoun": 0,
    "refVerb": 0,
    "sequence": 1,
    "talker": 99,
    "text": "He's been drinking Troll's sweat all night.  ",
    "verb": 1,
  },
  Object {
    "condition": 0,
    "index": 8,
    "noun": 8,
    "refCond": 166,
    "refNoun": 166,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "The front door is already barred.  ",
    "verb": 4,
  },
  Object {
    "condition": 0,
    "index": 9,
    "noun": 9,
    "refCond": 84,
    "refNoun": 65,
    "refVerb": 83,
    "sequence": 1,
    "talker": 99,
    "text": "The floor is strewn with flotsam and jetsam. ",
    "verb": 1,
  },
  Object {
    "condition": 2,
    "index": 10,
    "noun": 10,
    "refCond": 0,
    "refNoun": 0,
    "refVerb": 2,
    "sequence": 1,
    "talker": 99,
    "text": "You close and bar the door. ",
    "verb": 0,
  },
  Object {
    "condition": 3,
    "index": 11,
    "noun": 10,
    "refCond": 166,
    "refNoun": 119,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "\\"Oh, a wise guy, eh?\\" ",
    "verb": 0,
  },
  Object {
    "condition": 4,
    "index": 12,
    "noun": 10,
    "refCond": 167,
    "refNoun": 166,
    "refVerb": 53,
    "sequence": 1,
    "talker": 99,
    "text": "There's no need to knock this guy out; the Troll's Sweat he's been drinking has already done the job for you. ",
    "verb": 0,
  },
  Object {
    "condition": 5,
    "index": 13,
    "noun": 10,
    "refCond": 0,
    "refNoun": 194,
    "refVerb": 119,
    "sequence": 1,
    "talker": 99,
    "text": "You move the chair in front of the door. ",
    "verb": 0,
  },
  Object {
    "condition": 0,
    "index": 14,
    "noun": 10,
    "refCond": 119,
    "refNoun": 166,
    "refVerb": 194,
    "sequence": 1,
    "talker": 99,
    "text": "Wooden tables, benches, and chairs fill the space of this brigand meeting area.  A chandelier lights the room and  an unlit candelabra stands in the corner.  A rope hangs from the chandelier. ",
    "verb": 1,
  },
  Object {
    "condition": 0,
    "index": 15,
    "noun": 10,
    "refCond": 166,
    "refNoun": 166,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "None of the doors are locked.",
    "verb": 51,
  },
  Object {
    "condition": 0,
    "index": 16,
    "noun": 11,
    "refCond": 0,
    "refNoun": 0,
    "refVerb": 0,
    "sequence": 1,
    "talker": 99,
    "text": "Leave the rope alone for now.  It's serving a very important function by holding up the chandelier.",
    "verb": 4,
  },
  Object {
    "condition": 0,
    "index": 17,
    "noun": 11,
    "refCond": 119,
    "refNoun": 119,
    "refVerb": 149,
    "sequence": 1,
    "talker": 99,
    "text": "The rope holds up the chandelier.  It's used to lower the chandelier so that it can be lit. ",
    "verb": 1,
  },
  Object {
    "condition": 0,
    "index": 18,
    "noun": 12,
    "refCond": 166,
    "refNoun": 166,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "Most of the tables are grimy with dried beer and such.  The head table, though, is clean. ",
    "verb": 1,
  },
  Object {
    "condition": 0,
    "index": 19,
    "noun": 3,
    "refCond": 0,
    "refNoun": 0,
    "refVerb": 0,
    "sequence": 1,
    "talker": 99,
    "text": "You begin toppling the candelabra.  ",
    "verb": 4,
  },
  Object {
    "condition": 1,
    "index": 20,
    "noun": 3,
    "refCond": 166,
    "refNoun": 0,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "The candelabra has already fallen.  ",
    "verb": 4,
  },
  Object {
    "condition": 0,
    "index": 21,
    "noun": 3,
    "refCond": 166,
    "refNoun": 166,
    "refVerb": 166,
    "sequence": 1,
    "talker": 99,
    "text": "Sturdy-looking device for holding candles. ",
    "verb": 1,
  },
]
`;

exports[`pack-unpack test should unpack a message resource file 1`] = `
Object {
  "data": Object {
    "entries": Array [
      Object {
        "condition": 0,
        "noun": 1,
        "offset": 228,
        "refCond": 186,
        "refNoun": 20,
        "refSeq": 0,
        "refVerb": 0,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 0,
        "noun": 2,
        "offset": 244,
        "refCond": 166,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 0,
        "noun": 4,
        "offset": 289,
        "refCond": 166,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 0,
        "noun": 5,
        "offset": 334,
        "refCond": 0,
        "refNoun": 119,
        "refSeq": 0,
        "refVerb": 0,
        "sequence": 1,
        "talker": 99,
        "verb": 4,
      },
      Object {
        "condition": 0,
        "noun": 6,
        "offset": 364,
        "refCond": 119,
        "refNoun": 0,
        "refSeq": 0,
        "refVerb": 244,
        "sequence": 1,
        "talker": 99,
        "verb": 4,
      },
      Object {
        "condition": 1,
        "noun": 6,
        "offset": 442,
        "refCond": 166,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 4,
      },
      Object {
        "condition": 0,
        "noun": 6,
        "offset": 479,
        "refCond": 0,
        "refNoun": 0,
        "refSeq": 0,
        "refVerb": 0,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 0,
        "noun": 7,
        "offset": 518,
        "refCond": 0,
        "refNoun": 0,
        "refSeq": 0,
        "refVerb": 0,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 0,
        "noun": 8,
        "offset": 564,
        "refCond": 166,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 4,
      },
      Object {
        "condition": 0,
        "noun": 9,
        "offset": 600,
        "refCond": 84,
        "refNoun": 65,
        "refSeq": 0,
        "refVerb": 83,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 2,
        "noun": 10,
        "offset": 646,
        "refCond": 0,
        "refNoun": 0,
        "refSeq": 0,
        "refVerb": 2,
        "sequence": 1,
        "talker": 99,
        "verb": 0,
      },
      Object {
        "condition": 3,
        "noun": 10,
        "offset": 675,
        "refCond": 166,
        "refNoun": 119,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 0,
      },
      Object {
        "condition": 4,
        "noun": 10,
        "offset": 698,
        "refCond": 167,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 53,
        "sequence": 1,
        "talker": 99,
        "verb": 0,
      },
      Object {
        "condition": 5,
        "noun": 10,
        "offset": 809,
        "refCond": 0,
        "refNoun": 194,
        "refSeq": 0,
        "refVerb": 119,
        "sequence": 1,
        "talker": 99,
        "verb": 0,
      },
      Object {
        "condition": 0,
        "noun": 10,
        "offset": 851,
        "refCond": 119,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 194,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 0,
        "noun": 10,
        "offset": 1044,
        "refCond": 166,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 51,
      },
      Object {
        "condition": 0,
        "noun": 11,
        "offset": 1074,
        "refCond": 0,
        "refNoun": 0,
        "refSeq": 0,
        "refVerb": 0,
        "sequence": 1,
        "talker": 99,
        "verb": 4,
      },
      Object {
        "condition": 0,
        "noun": 11,
        "offset": 1174,
        "refCond": 119,
        "refNoun": 119,
        "refSeq": 0,
        "refVerb": 149,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 0,
        "noun": 12,
        "offset": 1267,
        "refCond": 166,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
      Object {
        "condition": 0,
        "noun": 3,
        "offset": 1358,
        "refCond": 0,
        "refNoun": 0,
        "refSeq": 0,
        "refVerb": 0,
        "sequence": 1,
        "talker": 99,
        "verb": 4,
      },
      Object {
        "condition": 1,
        "noun": 3,
        "offset": 1395,
        "refCond": 166,
        "refNoun": 0,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 4,
      },
      Object {
        "condition": 0,
        "noun": 3,
        "offset": 1432,
        "refCond": 166,
        "refNoun": 166,
        "refSeq": 0,
        "refVerb": 166,
        "sequence": 1,
        "talker": 99,
        "verb": 1,
      },
    ],
    "lastMsgNum": 1470,
    "length": 0,
    "nEntries": 22,
    "version": 218890383,
  },
  "strings": Array [
    "The back door. ",
    "They remind you of a bunch of Knochelkopfs. ",
    "There is a chandelier with a rope attached. ",
    "The door is already blocked. ",
    "No time for that again (although it WAS fun).  Just get the heck outta here. ",
    "The chandelier is out of your reach.",
    "Chandelier, used to hold up the rope. ",
    "He's been drinking Troll's sweat all night.  ",
    "The front door is already barred.  ",
    "The floor is strewn with flotsam and jetsam. ",
    "You close and bar the door. ",
    "\\"Oh, a wise guy, eh?\\" ",
    "There's no need to knock this guy out; the Troll's Sweat he's been drinking has already done the job for you. ",
    "You move the chair in front of the door. ",
    "Wooden tables, benches, and chairs fill the space of this brigand meeting area.  A chandelier lights the room and  an unlit candelabra stands in the corner.  A rope hangs from the chandelier. ",
    "None of the doors are locked.",
    "Leave the rope alone for now.  It's serving a very important function by holding up the chandelier.",
    "The rope holds up the chandelier.  It's used to lower the chandelier so that it can be lit. ",
    "Most of the tables are grimy with dried beer and such.  The head table, though, is clean. ",
    "You begin toppling the candelabra.  ",
    "The candelabra has already fallen.  ",
    "Sturdy-looking device for holding candles. ",
  ],
}
`;
